src/hooks/useGlobalState.tsx

```diff
@@ -1,25 +1,82 @@
-import React, { useContext, useReducer } from 'react'
+import React, { useContext, useReducer, useCallback, useRef, useMemo } from 'react'
 import { GlobalState } from 'reducers-actions'
 import { AnyAction } from 'combine-reducers'
+import { isEqual } from 'lodash'
+
+import { pendingOrdersAndLocalStorage, Middleware } from 'middlewares'
+import { compose } from 'utils'
 
 const GlobalStateContext = React.createContext({})
 
+const dispatchThatReturnsAction = (dispatch: React.Dispatch<AnyAction>) => (action: AnyAction): AnyAction => {
+  dispatch(action)
+  return action
+}
+
+const applyMiddlewares = (...middlewares: Middleware[]) => (
+  getState: () => GlobalState,
+  dispatch: React.Dispatch<AnyAction>,
+): React.Dispatch<AnyAction> => {
+  // fire off each middleware
+  const middlewareChain = middlewares
+    // ==============> e.g logger(getState, dispatch)
+    // returns ======> e.g loggerDispatch(next)(action)
+    .map(middleware => middleware(getState, dispatch))
+
+  const enhancedDispatch = compose(...middlewareChain)(dispatchThatReturnsAction(dispatch))
+
+  return enhancedDispatch
+}
+
+function useDeepCompareRef<R>(value?: R): R | undefined {
```

Comment:
To remove `R | undefined` ambiguity would be nice to use overloads:
```suggestion
function useDeepCompareRef<R>(value: R): R
function useDeepCompareRef(value: undefined): undefined
function useDeepCompareRef<R>(value?: R): R | undefined {
```
or something